vm/dispatcher: a test failure

https://github.com/google/syzkaller/actions/runs/10038063351/job/27739195812
```
panic: test timed out after 10m0s
running tests:
	TestPoolSplit (10m0s)

goroutine 45 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	/usr/local/go/src/time/sleep.go:176 +0x2d

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000102b60, {0x615e54?, 0x529cbc?}, 0x63e500)
	/usr/local/go/src/testing/testing.go:1649 +0x3c8
testing.runTests.func1(0x7c74c0?)
	/usr/local/go/src/testing/testing.go:2054 +0x3e
testing.tRunner(0xc000102b60, 0xc00017bc38)
	/usr/local/go/src/testing/testing.go:1595 +0xff
testing.runTests(0xc00011ebe0?, {0x7bd200, 0x2, 0x2}, {0xc00012ead0?, 0xc00017bcf8?, 0x7c6ca0?})
	/usr/local/go/src/testing/testing.go:2052 +0x445
testing.(*M).Run(0xc00011ebe0)
	/usr/local/go/src/testing/testing.go:1925 +0x636
main.main()
	_testmain.go:83 +0x1bf

goroutine 33 [sleep]:
time.Sleep(0x989680)
	/usr/local/go/src/runtime/time.go:195 +0x125
github.com/google/syzkaller/vm/dispatcher.(*testInstance).waitRun(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:156
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit(0xc000102d00?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:112 +0x52e
testing.tRunner(0xc0002[180](https://github.com/google/syzkaller/actions/runs/10038063351/job/27739195812#step:6:182)00, 0x63e500)
	/usr/local/go/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1648 +0x3ad

goroutine 34 [semacquire]:
sync.runtime_Semacquire(0xc0000c0080?)
	/usr/local/go/src/runtime/sema.go:62 +0x25
sync.(*WaitGroup).Wait(0x0?)
	/usr/local/go/src/sync/waitgroup.go:116 +0x48
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Loop(0x678b00, {0x676c70, 0xc000202280?})
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:70 +0x171
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func3()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:76 +0x2f
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:75 +0x292

goroutine 36 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 37 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 35 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 38 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 39 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 40 [chan receive]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:154
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x13b
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 41 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 42 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 43 [chan send]:
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Run(...)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:150
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:84 +0x125
created by github.com/google/syzkaller/vm/dispatcher.TestPoolSplit in goroutine 33
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:83 +0x2d8

goroutine 5 [chan send]:
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4.1({0x676c70, 0xc00014c500}, 0x18?, 0x5ef6c0?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:85 +0x34
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func4.(*Pool[...]).Run.func2(0xc00010e460?, 0x0?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:151 +0x38
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).runInstance(0x678b00, {0x676c70, 0xc000202280?}, 0xc0002[200](https://github.com/google/syzkaller/actions/runs/10038063351/job/27739195812#step:6:202)00?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:111 +0x5f7
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Loop.func1()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:65 +0x85
created by github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Loop in goroutine 34
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:63 +0xa5

goroutine 6 [select]:
github.com/google/syzkaller/vm/dispatcher.(*testInstance).run(0x40d785?, {0x676c70?, 0xc0002023c0?})
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:148 +0x6e
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func2({0x676c70?, 0xc0002023c0?}, 0xc000204140?, 0x619e65?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:68 +0x53
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).runInstance(0x678b00, {0x676c70, 0xc000202280?}, 0xc000220080?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:111 +0x5f7
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Loop.func1()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:65 +0x85
created by github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Loop in goroutine 34
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:63 +0xa5

goroutine 7 [select]:
make[1]: *** [Makefile:386: test] Error 1
make: *** [Makefile:315: presubmit_build] Error 2
github.com/google/syzkaller/vm/dispatcher.(*testInstance).run(0x40d785?, {0x676c70?, 0xc000202370?})
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:148 +0x6e
github.com/google/syzkaller/vm/dispatcher.TestPoolSplit.func2({0x676c70?, 0xc000202370?}, 0xc000204130?, 0x619e65?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool_test.go:68 +0x53
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).runInstance(0x678b00, {0x676c70, 0xc000202280?}, 0xc0002[201](https://github.com/google/syzkaller/actions/runs/10038063351/job/27739195812#step:6:203)00?)
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:111 +0x5f7
github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Loop.func1()
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:65 +0x85
created by github.com/google/syzkaller/vm/dispatcher.(*Pool[...]).Loop in goroutine 34
	/__w/syzkaller/syzkaller/gopath/src/github.com/google/syzkaller/vm/dispatcher/pool.go:63 +0xa5
FAIL	github.com/google/syzkaller/vm/dispatcher	600.010s
ok  	github.com/google/syzkaller/vm/isolated	0.376s	coverage: 0.5% of statements
ok  	github.com/google/syzkaller/vm/proxyapp	2.529s	coverage: 88.5% of statements
ok  	github.com/google/syzkaller/vm/vmimpl	0.376s	coverage: 15.0% of statements
FAIL
```

--- a/vm/dispatcher/pool_test.go
+++ b/vm/dispatcher/pool_test.go
@@ -79,17 +79,14 @@ func TestPoolSplit(t *testing.T) {
 
 	startedRuns := make(chan bool)
 	stopRuns := make(chan bool)
-	for i := 0; i < 10; i++ {
-		go func() {
-			mgr.Run(func(ctx context.Context, _ *testInstance, _ UpdateInfo) {
-				startedRuns <- true
-				select {
-				case <-ctx.Done():
-				case <-stopRuns:
-				}
-			})
-		}()
+	job := func(ctx context.Context, _ *testInstance, _ UpdateInfo) {
+		startedRuns <- true
+		select {
+		case <-ctx.Done():
+		case <-stopRuns:
+		}
 	}
+	go mgr.Run(job)
 
 	// So far, there are no reserved instances.
 	for i := 0; i < count; i++ {
@@ -113,9 +110,12 @@ func TestPoolSplit(t *testing.T) {
 	}
 	assert.EqualValues(t, 3, defaultCount.Load())
 
-	// Now let's finish all jobs.
+	// Now let's create and finish more jobs.
+	for i := 0; i < 10; i++ {
+		go mgr.Run(job)
+	}
 	mgr.ReserveForRun(2)
-	for i := 0; i < 9; i++ {
+	for i := 0; i < 10; i++ {
 		<-startedRuns
 		stopRuns <- true
 	}

